### PR TITLE
Test/mail

### PR DESCRIPTION
**Pull request by Totoual**
Created a test to test the aea/mail/base.py script

### CODE REVIEW

File: tests/test_mail.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	to_adr = []
+	msg = Message(content="hello")
+	message_bytes = ProtobufSerializer().encode(msg)
+	envelope = Envelope(to=to_adr, sender="Agent0", protocol_id="my_own_protocol", message=message_bytes)
+	assert (envelope, "Passing an empty receiver. Testing try/except by passing a list as sender")
+
+def test_inbox_empty():
+	my_queue = Queue() 
+	_inbox = InBox(my_queue)
+	assert(_inbox.empty(), "Must return an empty inbox")
```
Comment (DavidMinarsch) [inline tests/test_mail.py]: Comment should be relevant in the fail case, i.e. "Does not return an empty inbox."

File: tests/test_mail.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+from queue import Queue
+import pytest
+
+
+def test_envelope_initialisaztion() : 
+	msg = Message(content='hello')
+	message_bytes = ProtobufSerializer().encode(msg)
+	envelope = Envelope(to="Agent1", sender="Agent0", protocol_id="my_own_protocol", message=message_bytes)
+	
+	envelope.to = "Agent1"
```
Comment (DavidMinarsch) [inline tests/test_mail.py]: If you want to test the setters than maybe do:
assert envelope.to = "Agent1", "Cannot set to value on Envelope."

Similarly, for the others.

File: tests/test_mail.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	envelope.message = b"hello"
+	
+	assert (envelope,"Cannot generate a new envelope")
+
+def test_envelope_empty_receiver() : 
+	to_adr = []
+	msg = Message(content="hello")
+	message_bytes = ProtobufSerializer().encode(msg)
+	envelope = Envelope(to=to_adr, sender="Agent0", protocol_id="my_own_protocol", message=message_bytes)
+	assert (envelope, "Passing an empty receiver. Testing try/except by passing a list as sender")
```
Comment (DavidMinarsch) [inline tests/test_mail.py]: Same, this won't fail..

File: tests/test_mail.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	my_queue = Queue()
+	my_queue.put(message_bytes)
+	_inbox = InBox(my_queue)
+	assert(_inbox.get_nowait(), "Check for a message on the in queue and wait for no time.")
+
+#Test thet the outbox queue is empty
+def test_outbox_empty() :
+	my_queue = Queue()
+	_outbox = OutBox(my_queue)
+	assert(_outbox.empty(), "Checking if the outbox is empty")
```
Comment (DavidMinarsch) [inline tests/test_mail.py]: Write message for fail case, i.e." "Outbox is not empty."

File: aea/mail/base.py [hunk 5]
```diff
@@ -50,13 +50,12 @@ def __init__(self, to: Address,
...
-        assert type(self._to) == str or self._to is None
-        try:
+        try : 
+            assert type(self._to) == str or self._to is None
             if self._to is not None and type(self._to) == str:
                 self._to.encode('utf-8')
-        except Exception:
-            assert False
-
+        except AssertionError as error :
```
Comment (marcofavorito) [inline aea/mail/base.py]: the handling of the failing assertion is a good idea. However, we should decide if we tolerate this bad state or not. if we raise another exception or we just log a warning/error/exception message.

File: tests/test_mail.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	msg = Message(content='hello')
+	message_bytes = ProtobufSerializer().encode(msg)
+	envelope = Envelope(to="Agent1", sender="Agent0", protocol_id="my_own_protocol", message=message_bytes)
+	
+	envelope.to = "Agent1"
+	envelope.sender = "Agent0"
+	envelope.protocol_id = "my_own_protocol"
+	envelope.message = b"hello"
+	
+	assert (envelope,"Cannot generate a new envelope")
```
[the remark this answers is not in the document]
Comment (Totoual) [reply tests/test_mail.py]: Removed the parenthesis in order not to be always true.

File: tests/test_mail.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+#Test thet the outbox queue is empty
+def test_outbox_empty() :
+	my_queue = Queue()
+	_outbox = OutBox(my_queue)
+	assert _outbox.empty(), "Checking if the outbox is empty"
+
+
+''' Testing the mailBox()'''
+#It creates a new thread for the connection ( So interupt after passing all tests)
```
[the remark this answers is not in the document]
Comment (marcofavorito) [reply tests/test_mail.py]: @Totoual the same fix for the other functions.

File: tests/test_mail.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    assert envelope.sender == "ChangedSender", "Cannot set sender value on Envelope"
+    assert envelope.protocol_id == "my_changed_protocol", "Cannot set protocol_id on Envelope "
+    assert envelope.message == b"HelloWorld", "Cannot set message on Envelope"
+
+    assert envelope, "Cannot generate a new envelope"
+
+
+def test_envelope_empty_receiver():
+    """Tests what happens when we are passing wrong type of receiver."""
+    to_adr = []
```
Comment (DavidMinarsch) [inline tests/test_mail.py]: This should be = None

File: tests/test_mail.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    msg = Message(content="hello")
+    message_bytes = ProtobufSerializer().encode(msg)
+    envelope = Envelope(to=to_adr, sender="Agent0",
```
Comment (DavidMinarsch) [inline tests/test_mail.py]: then here assert it throws an error

**Pull request merged**